src/arch/cuda/hybridAPI/buddy_allocator.cpp

```diff
@@ -0,0 +1,187 @@
+#include "buddy_allocator.h"
+
+#include <cstdio>
+#include <cstdlib>
+#include <cmath>
+#include <algorithm>
+#include <cuda_runtime.h>
+
+namespace buddy {
+  void allocator::print_status() {
+    printf("(buckets)\n");
+    size_t free = 0;
+    for (int i = 0; i < bucket_count; i++) {
+      printf("bucket[%d]: ", i);
+      for (const auto& block : buckets[i]) {
+        free += block.size;
+        printf("{%p, %lu} ", block.ptr, block.size);
+      }
+      printf("\n");
+    }
+
+    printf("(alloc_map)\n");
+    size_t allocated = 0;
+    size_t used = 0;
+    for (const auto& elem : alloc_map) {
+      const auto& block = elem.second;
+      allocated += block.size;
+      used += block.requested;
+      printf("ptr: %p, size: %lu, req: %lu\n", elem.first, block.size, block.requested);
+    }
+
+    printf("(fragmentation) free: %lu, allocated: %lu, used: %lu\n", free, allocated, used);
+  }
+
+  size_t allocator::get_free_size() {
+    size_t free = 0;
+    for (int i = 0; i < bucket_count; i++) {
+      for (const auto& block : buckets[i]) {
+        free += block.size;
+      }
+    }
+
+    return free;
+  }
+
+  int allocator::get_bucket(size_t size) {
+    return (int)std::ceil(std::log2((double)size)) - 2;
+  }
+
+  int allocator::get_block_index(uint8_t* ptr, size_t size) {
+    return (size_t)(ptr - base_ptr) / size;
+  }
+
+  allocator::allocator(size_t size) : min_size(4), base_ptr(NULL) {
+    if (size <= 0) {
+      fprintf(stderr, "Allocator size has to be larger than 0 bytes\n");
+      abort();
+    }
+
+    // Request GPU memory (closest power of 2)
+    int total_size_log2 = std::ceil(std::log2((double)size));
+    total_size = (size_t)std::pow(2, total_size_log2);
+    cudaMalloc(&base_ptr, total_size);
```

Comment:
Should this check for an error?

Reply:
Good point, will add error checking for `cudaMalloc` and `cudaFree` calls